tempesta_fw/t/functional/testers/functional.py

```diff
@@ -44,24 +47,70 @@ def create_servers_helper(self, count, start_port=None, keep_alive=None,
                                                keep_alive=keep_alive,
                                                conns_n=connections))
 
+    def create_tester(self):
+        self.tester = deproxy.Deproxy(self.client, self.servers)
+
+    def configure_tester(self, message_chains):
+        self.tester.message_chains = message_chains
+
+    def create_tester_state(self, tester):
+        self.tester_state = control.DeproxyTester(tester)
+
     def setUp(self):
         self.client = None
+        self.client_state = None
         self.tempesta = None
         self.servers = []
         self.tester = None
+        self.tester_state = None
         tf_cfg.dbg(3) # Step to the next line after name of test case.
         tf_cfg.dbg(3, '\tInit test case...')
+        self.create_servers()
         self.create_tempesta()
+        self.create_client()
+        self.create_client_state(self.client)
+        self.create_tester()
+        self.create_tester_state(self.tester)
+
+    def force_stop(self):
+        # Close client connection before stopping the TempestaFW.
+        if self.client_state:
+            self.client_state.force_stop()
+
+        if self.tempesta:
+            self.tempesta.force_stop()
+
+        if self.tester_state:
+            self.tester_state.force_stop()
+
+        try:
+            control.finish_all_deproxy()
+        except:
+            print ('Unknown exception in stopping deproxy')
+
 
     def tearDown(self):
         # Close client connection before stopping the TempestaFW.
-        asyncore.close_all()
-        if self.client:
-            self.client.close()
+        if self.client_state:
+            self.client_state.stop()
+
         if self.tempesta:
             self.tempesta.stop()
-        if self.tester:
-            self.tester.close_all()
+
+        if self.tester_state:
+            self.tester_state.stop()
+
+        try:
+            control.finish_all_deproxy()
+        except:
+            print ('Unknown exception in stopping deproxy')
+
+        if self.client_state.state == control.STATE_ERROR:
```

Comment:
```
for proc in [client, server, ....]:
    if proc.state == STATE_ERROR:
        raise Exception("Error during stopping %s" % proc.__class__.__name__)
```
will do the trick without copy-paste. What do you think?

Reply:
It is possible

Reply:
fixed